pbj-core/pbj-compiler/src/main/java/com/hedera/pbj/compiler/impl/Common.java

```diff
@@ -172,14 +174,365 @@ public static String javaPrimitiveToObjectType(String primitiveFieldType) {
+       					"""
+						if ($fieldName != thatObj.$fieldName) {
+						    return false;
+						}
+						""".replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.FIXED64 ||
+						f.type() == Field.FieldType.INT64 ||
+						f.type() == Field.FieldType.SFIXED64 ||
+						f.type() == Field.FieldType.SINT64 ||
+						f.type() == Field.FieldType.UINT64) {
+					generatedCodeSoFar +=
+       					"""
+						if ($fieldName != thatObj.$fieldName) {
+						    return false;
+						}
+						""".replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.BOOL) {
+					generatedCodeSoFar +=
+       					"""
+						if ($fieldName != thatObj.$fieldName) {
+						    return false;
+						}
+						""".replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.DOUBLE) {
+				} else if (f.type() == Field.FieldType.FLOAT) {
+					generatedCodeSoFar +=
+       					"""
+						if ($fieldName != thatObj.$fieldName) {
+						    return false;
+						}
+						""".replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.DOUBLE) {
+					generatedCodeSoFar +=
+       					"""
+						if ($fieldName != thatObj.$fieldName) {
+						    return false;
+						}
+						""".replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.STRING ||
+						   f.type() == Field.FieldType.BYTES ||
+						   f.type() == Field.FieldType.ENUM ||
+				           f.parent() == null /* Process a sub-message */) {
+					generatedCodeSoFar += (
+       					"""
+						if ($fieldName == null && thatObj.$fieldName != null) {
+						    return false;
+						}
+						if ($fieldName != null && !$fieldName.equals(thatObj.$fieldName)) {
+						    return false;
+						}
+						""").replace("$fieldName", f.nameCamelFirstLower());
+				}
+				else {
+					throw new IllegalArgumentException("Unexpected field type for getting HashCode - " + f.type().toString());
+				}
+			}
+		}
+		return generatedCodeSoFar.indent(DEFAULT_INDENT);
+	}
+
+	/**
+	 * Get the equals codegen for a optional field.
+	 * @param generatedCodeSoFar The string that the codegen is generated into.
+	 * @param f The field for which to generate the equals code.
+	 * @return Updated codegen string.
+	 */
+	@NonNull
+	private static String getOptionalEqualsGeneration(String generatedCodeSoFar, Field f) {
+		switch (f.messageType()) {
+			case "StringValue" ->
+				generatedCodeSoFar += (
+                """
+                if (this.$fieldName == null && thatObj.$fieldName != null) {
+                    return false;
+                }
+                if (this.$fieldName != null && !$fieldName.equals(thatObj.$fieldName)) {
+				    return false;
+                }
+				""").replace("$fieldName", f.nameCamelFirstLower());
+			case "BoolValue" ->
+				generatedCodeSoFar += (
+                """
+                if ($fieldName instanceof Object) {
```

Comment:
When can a field not be an instance of Object?

Reply:
If it's a primitive.